File: python-glue/unified_ai/resilience/rate_limiter.py
"""Rate limiting implementation"""

import asyncio
import time
from typing import Optional
from dataclasses import dataclass, field
from threading import Lock


@dataclass
class TokenBucket:
    """Token bucket for rate limiting"""
    
    capacity: int
    refill_rate: float  # tokens per second
    tokens: float = field(init=False)
    last_refill: float = field(init=False)
    _lock: Lock = field(default_factory=Lock, init=False)
    
    def __post_init__(self):
        self.tokens = float(self.capacity)
        self.last_refill = time.time()
    
    def _refill(self) -> None:
        """Refill tokens based on elapsed time (must be called with lock held)"""
        now = time.time()
        elapsed = now - self.last_refill
        tokens_to_add = elapsed * self.refill_rate
        
        if tokens_to_add > 0:
            self.tokens = min(self.tokens + tokens_to_add, self.capacity)
            self.last_refill = now
    
    def try_acquire(self, tokens: int = 1) -> bool:
        """Try to acquire tokens"""
        with self._lock:
            self._refill()
            if self.tokens >= tokens:
                self.tokens -= tokens
                return True
            return False
    
    def wait_time(self) -> float:
        """Calculate wait time until tokens available"""
        with self._lock:
            self._refill()
            if self.tokens >= 1:
                return 0.0
            tokens_needed = 1.0 - self.tokens
            return tokens_needed / self.refill_rate


class RateLimiter:
    """Rate limiter using token bucket"""
    
    def __init__(self, capacity: int, refill_rate: float = None):
        """
        Initialize rate limiter
        
        Args:
            capacity: Maximum number of tokens (requests)
            refill_rate: Tokens per second (defaults to capacity/60 for per-minute rate)
        """
        if refill_rate is None:
            # Default: refill at rate that allows capacity requests per minute
            refill_rate = capacity / 60.0
        self.bucket = TokenBucket(capacity, refill_rate)
    
    def allow(self) -> bool:
        """Check if request is allowed (alias for try_acquire)"""
        return self.try_acquire(1)
    
    def remaining(self) -> int:
        """Get remaining tokens"""
        with self.bucket._lock:
            self.bucket._refill()
            return max(0, int(self.bucket.tokens))
    
    async def acquire(self, tokens: int = 1) -> None:
        """Acquire tokens, waiting if necessary"""
        while not self.bucket.try_acquire(tokens):
            wait_time = self.bucket.wait_time()
            if wait_time > 0:
                await asyncio.sleep(wait_time)
    
    def try_acquire(self, tokens: int = 1) -> bool:
        """Try to acquire tokens without waiting"""
        return self.bucket.try_acquire(tokens)

File: rust-core/src/resilience/rate_limiter.rs
use crate::error::{OrchestratorError, Result};
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

#[derive(Debug, Clone)]
pub struct TokenBucket {
    capacity: u32,
    tokens: u32,
    refill_rate: f64, // tokens per second
    last_refill: Instant,
}

impl TokenBucket {
    pub fn new(capacity: u32, refill_rate: f64) -> Self {
        Self {
            capacity,
            tokens: capacity,
            refill_rate,
            last_refill: Instant::now(),
        }
    }
    
    fn refill(&mut self) {
        let now = Instant::now();
        let elapsed = now.duration_since(self.last_refill).as_secs_f64();
        let tokens_to_add = (elapsed * self.refill_rate) as u32;
        
        if tokens_to_add > 0 {
            self.tokens = (self.tokens + tokens_to_add).min(self.capacity);
            self.last_refill = now;
        }
    }
    
    pub fn try_acquire(&mut self, tokens: u32) -> bool {
        self.refill();
        if self.tokens >= tokens {
            self.tokens -= tokens;
            true
        } else {
            false
        }
    }
    
    pub fn wait_time(&self) -> Duration {
        if self.tokens >= 1 {
            Duration::ZERO
        } else {
            let tokens_needed = 1.0 - self.tokens as f64;
            let seconds = tokens_needed / self.refill_rate;
            Duration::from_secs_f64(seconds)
        }
    }
}

#[derive(Debug, Clone)]
pub struct RateLimiter {
    bucket: Arc<Mutex<TokenBucket>>,
    name: String,
}

impl RateLimiter {
    pub fn new(name: impl Into<String>, capacity: u32, refill_rate: f64) -> Self {
        Self {
            bucket: Arc::new(Mutex::new(TokenBucket::new(capacity, refill_rate))),
            name: name.into(),
        }
    }
    
    pub async fn acquire(&self, tokens: u32) -> Result<()> {
        loop {
            let acquired = {
                let mut bucket = self.bucket.lock().unwrap();
                bucket.try_acquire(tokens)
            };
            
            if acquired {
                return Ok(());
            }
            
            let wait_time = {
                let bucket = self.bucket.lock().unwrap();
                bucket.wait_time()
            };
            
            if wait_time > Duration::ZERO {
                tokio::time::sleep(wait_time).await;
            }
        }
    }
    
    pub fn try_acquire(&self, tokens: u32) -> Result<()> {
        let mut bucket = self.bucket.lock().unwrap();
        if bucket.try_acquire(tokens) {
            Ok(())
        } else {
            Err(OrchestratorError::RateLimitExceeded(
                format!("Rate limit exceeded for {}", self.name)
            ))
        }
    }
}
